:book: Add pivoting documentation to user-guide

Rozzii opened the pull request:
This document is mostly the same as the old move.md documentation except a few minor modifications.

Rozzii commented:
/cc @fmuyassarov
/cc @furkatgofurov7

fmuyassarov commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -1,24 +1,21 @@
+## What is pivoting
 
-The following objects are moved during this process:
+CAPM3 implements support for CAPI's 'move/pivoting' feature.
```
```suggestion
Cluster-api-providfer-metal3 (CAPM3) implements support for CAPI's 'move/pivoting' feature.
```

fmuyassarov commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -1,24 +1,21 @@
-KubeadmConfig, Metal3MachineTemplate, Metal3IPPool, Metal3IPClaim, IPAddress,
-Metal3data, Metal3DataClaim, Metal3DataTemplate
+In Metal3, pivoting is performed by using the CAPI clusterctl tool provided by Cluster-API. clusterctl recognizes pivoting as move.
+During the pivot process clusterctl pauses any reconciliation of Cluster-API objects and this gets propagated to Cluster-api-provider-metal3
+(CAPM3) objects as well. Once all the objects are paused, the objects are created on the other side on the target cluster and deleted from the
```
```suggestion
During the pivot process clusterctl pauses any reconciliation of Cluster-API objects and this gets propagated to
CAPM3 objects as well. Once all the objects are paused, the objects are created on the other side on the target cluster and deleted from the
```

fmuyassarov commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -1,24 +1,21 @@
+## Prerequisits
 
-* **`clusterctl` is used to initialize both the bootstrap and target cluster.**
+* It is mandatory to use **`clusterctl` both the bootstrap and target cluster.**
```
This sentence seems a bit incomplete. Also
```suggestion
* It is mandatory to use `clusterctl` both the bootstrap and target cluster.
```

fmuyassarov reviewed (commented):
It is hard to comment here because some of the sentenced are not touched, and GitHub doesn't show them all. Can you maybe just delete the move.md and re-create a new so that it is possible for us to comment out ?

Rozzii commented:
> It is hard to comment here because some of the sentences are not touched, and GitHub doesn't show them all. Can you maybe just delete the move.md and re-create a new so that it is possible for us to comment out ?

I don't think I can, I have tried it but git is able to figure it out.

fmuyassarov commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -1,24 +1,21 @@
 
-**Main Objects:** Cluster, Machine, MachineSet, MachineDeployment,
-KubeadmControlplane, Metal3Cluster, Metal3Machine, and BareMetalHost
+CAPI Pivoting feature is a process for moving the provider components and declared Cluster API resources from a source
```
```suggestion
CAPI Pivoting feature is a process of moving the provider components and declared Cluster API resources from a source
```

Rozzii replied:
done

fmuyassarov commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -1,24 +1,21 @@
-Metal3data, Metal3DataClaim, Metal3DataTemplate
+In Metal3, pivoting is performed by using the CAPI clusterctl tool provided by Cluster-API. clusterctl recognizes pivoting as move.
+During the pivot process clusterctl pauses any reconciliation of Cluster-API objects and this gets propagated to (CAPM3) objects as well.
+Once all the objects are paused, the objects are created on the other side on the target cluster and deleted from the
+ephemeral cluster.
```
```suggestion
Once all the objects are paused, the objects are created on the other side on the target cluster and deleted from the
source cluster.
```
just to use the same terminology

Rozzii replied:
done

fmuyassarov commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -1,24 +1,21 @@
+## Prerequisits
 
-* **`clusterctl` is used to initialize both the bootstrap and target cluster.**
+* **It is mandatory to use `clusterctl` both the bootstrap and target cluster.**
```
This sentence seems missing something

Rozzii replied:
done

fmuyassarov commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -1,24 +1,21 @@
+* **It is mandatory to use `clusterctl` both the bootstrap and target cluster.**
 
   If the provider components are not installed using `clusterctl`, it will not
   be able to identify the objects to move.  Initializing the cluster using
```
Since it is hard to show what changes are requested, I will mark them in both.

So if the clusters are not initialized using clusterctl, all the CRDS of the objects to be moved to target cluster need to have these labels both **before performing move**. Note: **this** is not recommended, since the way clusterctl identifies objects to manage might change in future, so it’s always safe to install CRDs and controllers through **`clusterctl init`**.

Rozzii replied:
done

fmuyassarov commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -90,34 +87,34 @@ This can now be achieved with the following procedure:
 
-    `clusterctl init --infrastructure metal3:v0.3.1
---target-namespace metal3 --watching-namespace metal3`
+   `clusterctl init --infrastructure metal3:v1.1.
+   --target-namespace metal3 --watching-namespace metal3`
```
```suggestion
   `clusterctl init --infrastructure metal3:v1.1.1
   --target-namespace metal3 --watching-namespace metal3`
```

Rozzii replied:
done

fmuyassarov commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -90,34 +87,34 @@ This can now be achieved with the following procedure:
-  and CAPM3 as the infrastructure provider which also includes the Baremetal
-  Operator (BMO). All of the controllers will be installed on namespace `metal3`
-  and they will be watching over objects in namespace `metal3`.
+ This command will create the necessary CAPI controllers (CAPI, CABPK, CAKCP)
+ and CAPM3 as the infrastructure provider which also includes the Baremetal
+ Operator (BMO). All of the controllers will be installed on namespace `metal3`
+ and they will be watching over objects in namespace `metal3`.
```
```suggestion
 This command will create the necessary CAPI controllers (CAPI, CABPK, CAKCP)
 and CAPM3 as the infrastructure provider. All of the controllers will be installed on namespace `metal3`
 and they will be watching over objects in namespace `metal3`.
```
BMO is not part of CAPM3 anymore

Rozzii replied:
done

fmuyassarov commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -90,34 +87,34 @@ This can now be achieved with the following procedure:
-  directory.
+   In `metal3-dev-env` this is done through the scripts in `scripts/v1alphaX/`
+ directory.
 
 5. Wait for the target management cluster to be up and running and once it is up
-  get the kubeconfig for the new target management cluster.
+ get the kubeconfig for the new target management cluster.
 
 6. Use the new cluster’s kubeconfig to install the ironic-components in the
-  target cluster.
+ target cluster.
 
 7. Use clusterctl init with the new cluster’s kubeconfig to install the provider
-  components.
+ components.
```
I feel like we don't need to give examples with dev-env, but just generic instructions on how to perform move.

Rozzii replied:
done

Rozzii replied:
This has to stay like this because otherwise markdownlint has issue with it.

fmuyassarov reviewed (commented):
1. In some places clusterctl is marked as code block, while in other places as normal text. I don't mind whichever you choose, but it would be nice to follow the same style across the who page.

Rozzii commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -1,24 +1,21 @@
+## What is pivoting
 
-The following objects are moved during this process:
+Cluster-api-providfer-metal3 (CAPM3) implements support for CAPI's 'move/pivoting' feature.
```
yes it is intentional.

Xenwar commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,166 @@
+In Metal3, pivoting is performed by using the CAPI `clusterctl` tool provided by Cluster-API. `clusterctl` recognizes pivoting as move.
+During the pivot process `clusterctl` pauses any reconciliation of Cluster-API objects and this gets propagated to (CAPM3) objects as well.
+Once all the objects are paused, the objects are created on the other side on the target cluster and deleted from the
+source cluster.
```
Is the deletion always true ? and do we do it that way ?
@kashifest

kashifest replied:
Yes

Xenwar commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,166 @@
+
+  If the provider components are not installed using `clusterctl`, it will not
+  be able to identify the objects to move.  Initializing the cluster using
+  `clusterctl` essentially adds the following labels in the CRDs of each object
+  related.
```
```suggestion
  `clusterctl` essentially adds the following labels in the CRDs of each related object.
```

Rozzii replied:
done.

Xenwar commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,166 @@
+
+  So if the clusters are not initialized using `clusterctl`, all the CRDS of the
+  objects to be moved to target cluster needs to have these labels both in
+  bootstrap cluster and target clusteri befor performing the move. **Note**: This is not recommended, since
```
```suggestion
  bootstrap cluster and target cluster before performing the move. **Note**: This is not recommended, since
```

Rozzii replied:
done

Xenwar commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,166 @@
+  So if the clusters are not initialized using `clusterctl`, all the CRDS of the
+  objects to be moved to target cluster needs to have these labels both in
+  bootstrap cluster and target clusteri befor performing the move. **Note**: This is not recommended, since
+  the way `clusterctl` identifies objects to manage might change in future, so
```
```suggestion
  the way `clusterctl` identifies objects to manage might change in the future, so
```

Rozzii replied:
done

Xenwar commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,166 @@
+  objects to be moved to target cluster needs to have these labels both in
+  bootstrap cluster and target clusteri befor performing the move. **Note**: This is not recommended, since
+  the way `clusterctl` identifies objects to manage might change in future, so
+  it's always safe to install CRDs and controllers through `clusterctl` init.
```
```suggestion
  it's always safe to install CRDs and controllers through the `clusterctl init` sub-command.
```

Rozzii replied:
done

Xenwar commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,166 @@
+   undesirable.
+
+3. The provisioning bridge where the `ironic-endpoint-IP` is supposed to be
+   attached should have a static IP assignment on it before the ironic
```
```suggestion
   attached to should have a static IP assignment on it before the ironic
```

Rozzii replied:
done

Xenwar commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,166 @@
+  is to ensure that the Ironic Endpoint IP is also passed onto the target
+  cluster control plane. This also guarantees that once moving is done and the
+   management cluster is taken down, target cluster controlplane can re-claim
+   the ironic endpoint IP through keepalived. The end goal is to make ironic
```
```suggestion
   the Ironic endpoint IP through keepalived. The end goal is to make Ironic
```

Rozzii replied:
done

Xenwar commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,166 @@
+1. Create a temporary bootstrap cluster, e.g. using Kind or Minikube using
+ `clusterctl`.
+
+2. Install ironic  components namely: ironic, ironic-inspector, ironic-database,
```
```suggestion
2. Install Ironic  components namely: ironic, ironic-inspector, ironic-database,
```

Rozzii replied:
done

Xenwar commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,166 @@
+
+1. The move process should be done when the BMHs are in a steady state. BMHs
+   should not be moved while any operation is on-going i.e. BMH is in provisioning
+   state. This will result in failure since the interaction between IPA and ironic
```
```suggestion
   state. This will result in failure since the interaction between IPA and Ironic
```

Rozzii replied:
done

Xenwar commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,166 @@
+   pod/containers start to operate in the target cluster. This is important since
+   `ironic-endpoint-keepalived` container will only assign the `ironic-endpoint-IP`
+   on the provisioning bridge in target cluster when it has an IP on it. Otherwise
+   it will fail to attach the IP and ironic will be unreachable. This is crucial
```
```suggestion
   it will fail to attach the IP and Ironic will be unreachable. This is crucial
```

Rozzii replied:
done

Xenwar commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,166 @@
+1. The move process should be done when the BMHs are in a steady state. BMHs
+   should not be moved while any operation is on-going i.e. BMH is in provisioning
+   state. This will result in failure since the interaction between IPA and ironic
+   gets broken and as a result ironic's database might not be repopulated and
```
```suggestion
   gets broken and as a result Ironic's database might not be repopulated and
```

Rozzii replied:
done

Xenwar commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,166 @@
+   the BMH might change after the move and the DHCP-leases from the management
+   cluster are not moved to target cluster.
+
+2. Before the move process is initialized, it is important to delete the ironic
+   pod/ironic containers. If ironic is deployed in cluster the deployment is named
+   `metal3-ironic`, if it is deployed locally outside the cluster, the containers
+   to be deleted are `ironic`, `ironic-inspector`, `ironic-endpoint-keepalived`,
+   `ironic-dnsmasq`, `ironic-httpd` and `mariadb`. If ironic is not deleted
+   before move, the old ironic might interfere with the operations of
+   the new ironic deployed in target cluster since the database of the first ironic
+   instance is not cleaned when the BMHs are moved. Also there would be two dnsmasq
+   existent in the deployment if there would be two ironic deployment which is
+   undesirable.
```
```suggestion
2. Before the move process is initialized, it is important to delete the Ironic
   pod/ironic containers. If Ironic is deployed in cluster the deployment is named
   `metal3-ironic`, if it is deployed locally outside the cluster, the containers
   to be deleted are `ironic`, `ironic-inspector`, `ironic-endpoint-keepalived`,
   `ironic-dnsmasq`, `ironic-httpd` and `mariadb`. If Ironic is not deleted
   before move, the old ironic might interfere with the operations of
   the new ironic deployed in target cluster since the database of the first Ironic
   instance is not cleaned when the BMHs are moved. Also there would be two dnsmasq
   existent in the deployment if there would be two ironic deployment which is
   undesirable.
```

Rozzii replied:
done

Xenwar commented:
/lgtm

kashifest commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,167 @@
+
+CAPI Pivoting feature is a process of moving the provider components and declared Cluster API resources from a source
+management cluster to a target management cluster by using the `clusterctl` functionality called "move".
+More information about the general CAPI "move" functionality can be find [here](https://cluster-api.sigs.k8s.io/clusterctl/commands/move.html).
```
```suggestion
More information about the general CAPI "move" functionality can be found [here](https://cluster-api.sigs.k8s.io/clusterctl/commands/move.html).
```

kashifest commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,167 @@
+  Baremetal machines boot over a network with a DHCP server. This requires
+  maintaining a fixed IP end points towards the provisioning network. This is
+  achieved through _keepalived_. A new container is added namely
+  _ironic-endpoint-keepalived_ which maintains the Ironic Endpoint using
```
```suggestion
  _ironic-endpoint-keepalived_ in the ironic deployment which maintains the Ironic Endpoint using
```

kashifest commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,167 @@
+This can now be achieved with the following procedure:
+
+1. Create a temporary bootstrap cluster, e.g. using Kind or Minikube using
+ `clusterctl`.
```
kind or minikube is not setup using clusterctl. Rather the provider components in kind or minikube are. This line needs to be tweaked a little bit to phrase this appropriately.

kashifest commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,167 @@
+ `clusterctl`.
+
+2. Install Ironic  components namely: ironic, ironic-inspector, ironic-database,
+   ironic-endpoint-keepalived, httpd and dnsmasq.
```
ironic-database? do we have a container with that name ?

kashifest commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,167 @@
+
+    Example:
+
+    `clusterctl init --kubeconfig target.yaml --infrastructure metal3:v0.3.1
```
Can you keep the same version for metal3 in the example as in step 3

kashifest commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,167 @@
+   pod/Ironic containers. If Ironic is deployed in cluster the deployment is named
+   `metal3-ironic`, if it is deployed locally outside the cluster, the containers
+   to be deleted are `ironic`, `ironic-inspector`, `ironic-endpoint-keepalived`,
+   `ironic-dnsmasq`, `ironic-httpd` and `mariadb`. If Ironic is not deleted
```
is mariadb still deployed locally?

Rozzii replied:
I think I will remove the text related to running ironic locally as that is totally up to the user how the ironic is configured in that case.

kashifest reviewed (commented):
Few nits inline. Good work!

furkatgofurov7 commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,167 @@
+
+## What is pivoting
+
+Cluster-api-provider-metal3 (CAPM3) implements support for CAPI's 'move/pivoting' feature.
```
```suggestion
Cluster API Provider Metal3 (CAPM3) implements support for CAPI's 'move/pivoting' feature.
```

furkatgofurov7 commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,167 @@
+Once all the objects are paused, the objects are created on the other side on the target cluster and deleted from the
+source cluster.
+
+## Prerequisits
```
```suggestion
## Prerequisite
```

furkatgofurov7 commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,167 @@
+* **BMH objects have correct status annotation.**
+
+  Since BMH _status_ holds important information regarding the BMH itself, BMH with
+  status has to be moved and BMH has to be reconstructed with correct status in
```
```suggestion
  status has to be moved and it has to be reconstructed with correct status in
```

furkatgofurov7 commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,167 @@
+  Since BMH _status_ holds important information regarding the BMH itself, BMH with
+  status has to be moved and BMH has to be reconstructed with correct status in
+  target  cluster before it is being reconciled. This is now done through BMH
+  status annotation in BMO.
```
you could provide a reference to BMO doc here

Rozzii replied:
I would rather have somebody add it later when the bmo documentation is part of the user-guide.

furkatgofurov7 replied:
Is there an open issue to add it to the user guide? If not, why not to reference it

Rozzii replied:
Yes there is https://github.com/metal3-io/metal3-docs/issues/221, but based on the PR https://github.com/metal3-io/metal3-docs/pull/237 I can preemptively place the link, I hope the file names won't change.

furkatgofurov7 commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,167 @@
+  keepalived. The motivation behind maintaining Ironic Endpoint with Keepalived
+  is to ensure that the Ironic Endpoint IP is also passed onto the target
+  cluster control plane. This also guarantees that once moving is done and the
+   management cluster is taken down, target cluster controlplane can re-claim
+   the Ironic endpoint IP through keepalived. The end goal is to make Ironic
+   endpoint reachable in the target cluster.
```
these lines shifted to the right for some reason

furkatgofurov7 commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,167 @@
+
+  If not, it has to be deployed before the `clusterctl init` and the BMH CRDs
+  need to be labeled accordingly manually. This is a prerequisite for both the
+  management and the target cluster.
```
You could mention the CAPM3 API version which deploys BMO as part of it and the ones does not to make it more clear for the user

furkatgofurov7 commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,167 @@
+objects to target cluster can be described as follows:
+
+The move process can be bounded with the creation of a temporary bootstrap
+cluster used to provision a target Management cluster.
```
```suggestion
cluster used to provision a target management cluster.
```

furkatgofurov7 commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,167 @@
+
+   Example:
+
+   `clusterctl init --infrastructure metal3:v1.1.1
```
we do not have the CAPM3 v1.1.1 release yet

fmuyassarov commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,170 @@
+
+3. Use clusterctl init to install the provider components
+
+   Example:
+
+   `clusterctl init --infrastructure metal3:v1.1.0
+   --target-namespace metal3 --watching-namespace metal3`
```
```suggestion
   ```bash
   clusterctl init --infrastructure metal3:v1.1.0
   --target-namespace metal3 --watching-namespace metal3
   ```
```

fmuyassarov commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,170 @@
+   `clusterctl init --infrastructure metal3:v1.1.0
+   --target-namespace metal3 --watching-namespace metal3`
+
+ This command will create the necessary CAPI controllers (CAPI, CABPK, CAKCP)
+ and CAPM3 as the infrastructure provider. All of the controllers will be installed
+ on namespace `metal3` and they will be watching over objects in namespace `metal3`.
```
There is indentation issue if check https://deploy-preview-254--metal3io.netlify.app/capm3/pivoting.html
```suggestion
     This command will create the necessary CAPI controllers (CAPI, CABPK, CAKCP)
     and CAPM3 as the infrastructure provider. All of the controllers will be installed
     on namespace `metal3` and they will be watching over objects in namespace `metal3`.
```

fmuyassarov commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,170 @@
+6. Use the new cluster’s kubeconfig to install the ironic-components in the
+ target cluster.
+
+7. Use `clusterctl` init with the new cluster’s kubeconfig to install the provider
+ components.
+
+    Example:
+
+    `clusterctl init --kubeconfig target.yaml --infrastructure metal3:v1.1.0
+  --target-namespace metal3 --watching-namespace metal3`
+
```
```suggestion
7. Use `clusterctl` init with the new cluster’s kubeconfig to install the provider
 components.
    ```bash
     clusterctl init --kubeconfig target.yaml --infrastructure metal3:v1.1.0
    --target-namespace metal3 --watching-namespace metal3
    ```
```

fmuyassarov commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,170 @@
+    `clusterctl init --kubeconfig target.yaml --infrastructure metal3:v1.1.0
+  --target-namespace metal3 --watching-namespace metal3`
+
+8. Use `clusterctl` move to move the Cluster API resources from the bootstrap
+ cluster to the target management cluster.
+
+    Example:
+
+    `clusterctl move --to-kubeconfig target.yaml -n metal3 -v 10`
+
```
```suggestion
8. Use `clusterctl` move to move the Cluster API resources from the bootstrap
 cluster to the target management cluster.
 
    ```bash
    clusterctl move --to-kubeconfig target.yaml -n metal3 -v 10
    ```

```

fmuyassarov commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,170 @@
+
+## What is pivoting
+
+Cluster Api Provider Metal3 (CAPM3) implements support for CAPI's 'move/pivoting' feature.
```
```suggestion
Cluster API Provider Metal3 (CAPM3) implements support for Cluster API (CAPI)'s 'move/pivoting' feature.
```

fmuyassarov commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,170 @@
+
+* **BMH objects have correct status annotation.**
+
+  Since BMH _status_ holds important information regarding the BMH itself, BMH with
```
```suggestion
  Since BareMetalHost (BMH) _status_ holds important information regarding the BMH itself, BMH with
```

fmuyassarov commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,170 @@
+  the way `clusterctl` identifies objects to manage might change in the future, so
+  it's always safe to install CRDs and controllers through the `clusterctl init` sub-command.
+
+* **BMH objects have correct status annotation.**
```
```suggestion
* **BareMetalHost objects have correct status annotation.**
```

fmuyassarov commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,170 @@
+  Since BMH _status_ holds important information regarding the BMH itself, BMH with
+  status has to be moved and it has to be reconstructed with correct status in
+  target  cluster before it is being reconciled. This is now done through BMH
+  status annotation in [BMO](../bmo/introduction.md).
```
```suggestion
  [status annotation](https://github.com/metal3-io/baremetal-operator/blob/main/docs/statusAnnotation.md).
```

Rozzii replied:
I think it would be better if we would reference the BMO documentation of the same book, that will also contain examples of the BMH CRDs.

fmuyassarov replied:
agreed

fmuyassarov commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,170 @@
+
+* **BMO is deployed as part of CAPM3.**
+
+  If not, it has to be deployed before the `clusterctl init` and the BMH CRDs
+  need to be labeled accordingly manually. Separate labeling for BMH CRDs is required
+  because since CAPM3 version [v0.5.0](https://github.com/metal3-io/cluster-api-provider-metal3/releases/tag/v0.5.0)
+  BMO/BMH CRDs are not deplopyed as part of CAPM3 deployment anymore.
+  This is a prerequisite for both the management and the target cluster.
```
```suggestion
  If not, it has to be deployed before the `clusterctl init` and the BMH CRDs
  need to be labeled accordingly manually. Separate labeling for BMH CRDs is required
  because since CAPM3 release [v0.5.0](https://github.com/metal3-io/cluster-api-provider-metal3/releases/tag/v0.5.0)
  BMO/BMH CRDs are not deplopyed as part of CAPM3 deployment anymore.
  This is a prerequisite for both the management and the target cluster.
```
Can we also mention what labels need to be applied?

Rozzii replied:
It has been mentioned between line 24 to 28

fmuyassarov reviewed (commented):
This is looks great @Rozzii and thank a lot for fixing all the comments. I have last round of comments in line and here, otherwise it is in a good shape. Also, I don't think attaching a video here is a good idea. Because, very soon we are expecting a change in the pivoting workflow, so the workflow will be outdated. Apart from that, it is hard to maintain the video in case there are some changes. Maybe we can upload it in the Metal3 YouTube channel, and add a reference link here?

Also, we are mixing the usage of terms bootstrap, source and ephemeral. Can we perhaps stick to one of them across the whole document?

Important notes section seems as something that can we added under prerequisites section. What do you think if we move it up ?

Rozzii commented:
> This is looks great @Rozzii and thank a lot for fixing all the comments. I have last round of comments in line and here, otherwise it is in a good shape. Also, I don't think attaching a video here is a good idea. Because, very soon we are expecting a change in the pivoting workflow, so the workflow will be outdated. Apart from that, it is hard to maintain the video in case there are some changes. Maybe we can upload it in the Metal3 YouTube channel, and add a reference link here?
> 
> Also, we are mixing the usage of terms bootstrap, source and ephemeral. Can we perhaps stick to one of them across the whole document?
> 
> Important notes section seems as something that can we added under prerequisites section. What do you think if we move it up ?

I have unified the the source-ephemeral-bootstrap names to bootstrap, I agree I don't like the gif either so I have removed it, and sure in the future I agree that we should have a video and upload it to the YouTube channel.
I have moved the ##Important notes section up .

fmuyassarov commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,172 @@
+
+  **Note**: This is not recommended, since
+  the way `clusterctl` identifies objects to manage might change in the future, so
+  it's always safe to install CRDs and controllers through the `clusterctl init` sub-command.
+
+* **BareMetalHost objects have correct status annotation.**
+
+  Since BareMetalHost (BMH) _status_ holds important information regarding the BMH itself, BMH with
+  status has to be moved and it has to be reconstructed with correct status in
+  target  cluster before it is being reconciled. This is now done through BMH
+  status annotation in [BMO](../bmo/introduction.md).
+
+* **Maintain connectivity towards provisioning network.**
+
+  Baremetal machines boot over a network with a DHCP server. This requires
+  maintaining a fixed IP end points towards the provisioning network. This is
+  achieved through _keepalived_. A new container is added namely
+  _ironic-endpoint-keepalived_ in the ironic deployment which maintains the Ironic Endpoint using
+  keepalived. The motivation behind maintaining Ironic Endpoint with Keepalived
+  is to ensure that the Ironic Endpoint IP is also passed onto the target
+  cluster control plane. This also guarantees that once moving is done and the
+  management cluster is taken down, target cluster controlplane can re-claim
+  the Ironic endpoint IP through keepalived. The end goal is to make Ironic
+  endpoint reachable in the target cluster.
+
+* **BMO is deployed as part of CAPM3.**
+
+  If not, it has to be deployed before the `clusterctl init` and the BMH CRDs
+  need to be labeled accordingly manually. Separate labeling for BMH CRDs is required
+  because since CAPM3 release [v0.5.0](https://github.com/metal3-io/cluster-api-provider-metal3/releases/tag/v0.5.0)
+  BMO/BMH CRDs are not deplopyed as part of CAPM3 deployment anymore.
+  This is a prerequisite for both the management and the target cluster.
+
+* **Objects should have a proper owner reference chain.**
+
+  `clusterctl move` moves all the objects to the target cluster following the
+[owner reference chain](https://cluster-api.sigs.k8s.io/clusterctl/provider-contract.html#ownerreferences-chain).
+So, it is necessary to verify that all the desired objects that needs to
+be moved to the target cluster have a proper owner reference chain.
+
```
```suggestion
1. **It is mandatory to use `clusterctl` for both the bootstrap and target cluster**

      If the provider components are not installed using `clusterctl`, it will not
      be able to identify the objects to move.  Initializing the cluster using
      `clusterctl` essentially adds the following labels in the CRDs of each related object.

      ```yaml
      labels:
      - clusterctl.cluster.x-k8s.io: ""
      - cluster.x-k8s.io/provider: "<provider-name>"
      ```

      So if the clusters are not initialized using `clusterctl`, all the CRDS of the
      objects to be moved to target cluster needs to have these labels both in
      bootstrap cluster and target cluster before performing the move.

      **Note**: This is not recommended, since
      the way `clusterctl` identifies objects to manage might change in the future, so
      it's always safe to install CRDs and controllers through the `clusterctl init` sub-command.

1. **BareMetalHost objects have correct status annotation**

   Since BareMetalHost (BMH) _status_ holds important information regarding the BMH itself, BMH with
   status has to be moved and it has to be reconstructed with correct status in
   target  cluster before it is being reconciled. This is now done through BMH
   status annotation in [BMO](../bmo/introduction.md).

1. **Maintain connectivity towards provisioning network**

   Baremetal machines boot over a network with a DHCP server. This requires
   maintaining a fixed IP end points towards the provisioning network. This is
   achieved through _keepalived_. A new container is added namely
   _ironic-endpoint-keepalived_ in the ironic deployment which maintains the Ironic Endpoint using
   keepalived. The motivation behind maintaining Ironic Endpoint with Keepalived
   is to ensure that the Ironic Endpoint IP is also passed onto the target
   cluster control plane. This also guarantees that once moving is done and the
   management cluster is taken down, target cluster controlplane can re-claim
   the Ironic endpoint IP through keepalived. The end goal is to make Ironic
   endpoint reachable in the target cluster.

1. **BMO is deployed as part of CAPM3**

   If not, it has to be deployed before the `clusterctl init` and the BMH CRDs
   need to be labeled accordingly manually. Separate labeling for BMH CRDs is required
   because since CAPM3 release [v0.5.0](https://github.com/metal3-io/cluster-api-provider-metal3/releases/tag/v0.5.0)
   BMO/BMH CRDs are not deplopyed as part of CAPM3 deployment anymore.
   This is a prerequisite for both the management and the target cluster.

1. **Objects should have a proper owner reference chain**

   `clusterctl move` moves all the objects to the target cluster following the
   [owner reference chain](https://cluster-api.sigs.k8s.io/clusterctl/provider-contract.html#ownerreferences-chain).
   So, it is necessary to verify that all the desired objects that needs to
   be moved to the target cluster have a proper owner reference chain.
```

fmuyassarov replied:
so that it looks like
![pre](https://user-images.githubusercontent.com/35802557/159922022-588ac489-4280-4c83-820a-b22134f27991.png)

fmuyassarov replied:
and probably should be **Prerequisites** (in plural)

Rozzii replied:
It was originally like that but then I was informed to change it to Prerequisite

fmuyassarov commented on docs/user-guide/src/capm3/pivoting.md:
```diff
@@ -0,0 +1,172 @@
+
+  **Note**: This is not recommended, since
+  the way `clusterctl` identifies objects to manage might change in the future, so
+  it's always safe to install CRDs and controllers through the `clusterctl init` sub-command.
+
+* **BareMetalHost objects have correct status annotation.**
+
+  Since BareMetalHost (BMH) _status_ holds important information regarding the BMH itself, BMH with
+  status has to be moved and it has to be reconstructed with correct status in
+  target  cluster before it is being reconciled. This is now done through BMH
+  status annotation in [BMO](../bmo/introduction.md).
+
+* **Maintain connectivity towards provisioning network.**
+
+  Baremetal machines boot over a network with a DHCP server. This requires
+  maintaining a fixed IP end points towards the provisioning network. This is
+  achieved through _keepalived_. A new container is added namely
+  _ironic-endpoint-keepalived_ in the ironic deployment which maintains the Ironic Endpoint using
+  keepalived. The motivation behind maintaining Ironic Endpoint with Keepalived
+  is to ensure that the Ironic Endpoint IP is also passed onto the target
+  cluster control plane. This also guarantees that once moving is done and the
+  management cluster is taken down, target cluster controlplane can re-claim
+  the Ironic endpoint IP through keepalived. The end goal is to make Ironic
+  endpoint reachable in the target cluster.
+
+* **BMO is deployed as part of CAPM3.**
+
+  If not, it has to be deployed before the `clusterctl init` and the BMH CRDs
+  need to be labeled accordingly manually. Separate labeling for BMH CRDs is required
+  because since CAPM3 release [v0.5.0](https://github.com/metal3-io/cluster-api-provider-metal3/releases/tag/v0.5.0)
+  BMO/BMH CRDs are not deplopyed as part of CAPM3 deployment anymore.
+  This is a prerequisite for both the management and the target cluster.
+
+* **Objects should have a proper owner reference chain.**
+
+  `clusterctl move` moves all the objects to the target cluster following the
+[owner reference chain](https://cluster-api.sigs.k8s.io/clusterctl/provider-contract.html#ownerreferences-chain).
+So, it is necessary to verify that all the desired objects that needs to
+be moved to the target cluster have a proper owner reference chain.
+
```
```suggestion
1. **It is mandatory to use `clusterctl` for both the bootstrap and target cluster**

      If the provider components are not installed using `clusterctl`, it will not
      be able to identify the objects to move.  Initializing the cluster using
      `clusterctl` essentially adds the following labels in the CRDs of each related object.

      ```yaml
      labels:
      - clusterctl.cluster.x-k8s.io: ""
      - cluster.x-k8s.io/provider: "<provider-name>"
      ```

      So if the clusters are not initialized using `clusterctl`, all the CRDS of the
      objects to be moved to target cluster needs to have these labels both in
      bootstrap cluster and target cluster before performing the move.

      **Note**: This is not recommended, since
      the way `clusterctl` identifies objects to manage might change in the future, so
      it's always safe to install CRDs and controllers through the `clusterctl init` sub-command.

1. **BareMetalHost objects have correct status annotation**

   Since BareMetalHost (BMH) _status_ holds important information regarding the BMH itself, BMH with
   status has to be moved and it has to be reconstructed with correct status in
   target  cluster before it is being reconciled. This is now done through BMH
   status annotation in [BMO](../bmo/introduction.md).

1. **Maintain connectivity towards provisioning network**

   Baremetal machines boot over a network with a DHCP server. This requires
   maintaining a fixed IP end points towards the provisioning network. This is
   achieved through _keepalived_. A new container is added namely
   _ironic-endpoint-keepalived_ in the ironic deployment which maintains the Ironic Endpoint using
   keepalived. The motivation behind maintaining Ironic Endpoint with Keepalived
   is to ensure that the Ironic Endpoint IP is also passed onto the target
   cluster control plane. This also guarantees that once moving is done and the
   management cluster is taken down, target cluster controlplane can re-claim
   the Ironic endpoint IP through keepalived. The end goal is to make Ironic
   endpoint reachable in the target cluster.

1. **BMO is deployed as part of CAPM3**

   If not, it has to be deployed before the `clusterctl init` and the BMH CRDs
   need to be labeled accordingly manually. Separate labeling for BMH CRDs is required
   because since CAPM3 release [v0.5.0](https://github.com/metal3-io/cluster-api-provider-metal3/releases/tag/v0.5.0)
   BMO/BMH CRDs are not deplopyed as part of CAPM3 deployment anymore.
   This is a prerequisite for both the management and the target cluster.

1. **Objects should have a proper owner reference chain**

   `clusterctl move` moves all the objects to the target cluster following the
   [owner reference chain](https://cluster-api.sigs.k8s.io/clusterctl/provider-contract.html#ownerreferences-chain).
   So, it is necessary to verify that all the desired objects that needs to
   be moved to the target cluster have a proper owner reference chain.
```

fmuyassarov reviewed (commented):
/approve

fmuyassarov commented:
Thanks this looks good to me.
/cc @kashifest @furkatgofurov7 @namnx228

kashifest commented:
/lgtm
